impl/src/main/java/com/sun/faces/facelets/tag/faces/IterationIdManager.java

```diff
@@ -81,9 +81,14 @@ static boolean isIterating(FaceletContext context) {
     }
 
     private static Deque<Set<String>> _getStackOfTrackedIds(FaceletContext ctx) {
-        Deque<Set<String>> stack = (Deque<Set<String>>) ctx.getAttribute(_STACK_OF_TRACKED_IDS);
+        Deque<Set<String>> stack = (Deque<Set<String>>)ctx.getAttribute(_STACK_OF_TRACKED_IDS);
         if (stack == null) {
-            stack = new LinkedList<>();
+            synchronized(IterationIdManager.class) {
+                if(stack == null) {
```

Comment:
How can `stack` become non-null here?

Reply:
See https://www.baeldung.com/java-singleton-double-checked-locking

Reply:
But `stack` is local variable.

Reply:
You are right. Then the problem is using it out of the context.  We need a synchronized list and synchronize every access to this stack. 

We might have to synchronize the whole method. Because two threads are initializing and stuffing the list into the context overwriting the other

Reply:
Actually I wonder if I have to check the context get again inside the synchronize and not the local list

Reply:
OK @pzygielo I ammended my commit.  Can you check again?  I know look at the stack from the Session again and if its still NULL I create the new list